src/backend/utils/adt/network.c

```diff
@@ -389,6 +416,415 @@ network_cmp(PG_FUNCTION_ARGS)
+ *        bits are equal, so we're in effect only comparing subnet bits).
+ *
+ * When generating abbreviated keys for SortSupport, we pack in as much
+ * information as we have remove for that will ensure that when comparing those
+ * keys as integers, the rules above will be respected.
+ *
+ * In most cases, that means just the most significant of the netmasked bits
+ * are retained, but in the case of IPv4 addresses on 64-bit architecture, we
+ * can hold almost all relevant information for comparisons to the degree that
+ * we'd almost never have to fall back to non-abbreviated comparison. In all
+ * cases, there should be enough data present for the abbreviated keys to
+ * perform very well in all except the very worst of circumstances (and in
+ * those, key abbreviation will probably be aborted as we detect low
+ * cardinality).
+ *
+ * IPv4
+ * ----
+ *
+ * 32-bit architecture:
+ *
+ * Start with 1 bit for the IP family (IPv4 or IPv6; this bit is present in
+ * every case below) followed by all but 1 of the netmasked bits. If those are
+ * all equal, the system will have to fall back to non-abbreviated comparison.
+ *
+ * +----------+---------------------+
+ * | 1 bit IP |   31 bits netmask   |     (up to 1 bit
+ * |  family  |     (truncated)     |   netmask omitted)
+ * +----------+---------------------+
+ *
+ * 64-bit architecture:
+ *
+ * The most interesting case: we have space to store all netmasked bits,
+ * followed by the netmask size, followed by 25 bits of the subnet. We lay the
+ * bits out carefully this way so that the entire value can be compared as an
+ * integer while still adhering to the inet/cidr sorting rules stated above.
+ *
+ * 25 bits is more than enough to store most subnets (all but /6 or smaller),
+ * so in the vast majority of cases these keys should hold enough information
+ * to save trips to the heap.
+ *
+ * +----------+-----------------------+--------------+--------------------+
+ * | 1 bit IP |    32 bits netmask    |    6 bits    |   25 bits subnet   |
+ * |  family  |        (full)         | netmask size |    (truncated)     |
+ * +----------+-----------------------+--------------+--------------------+
+ *
+ * IPv6
+ * ----
+ *
+ * 32-bit architecture:
+ *
+ * +----------+---------------------+
+ * | 1 bit IP |   31 bits netmask   |    (up to 97 bits
+ * |  family  |     (truncated)     |   netmask omitted)
+ * +----------+---------------------+
+ *
+ * 64-bit architecture:
+ *
+ * +----------+---------------------------------+
+ * | 1 bit IP |         63 bits netmask         |    (up to 65 bits
+ * |  family  |           (truncated)           |   netmask omitted)
+ * +----------+---------------------------------+
+ *
+ */
+static Datum
+network_abbrev_convert(Datum original, SortSupport ssup)
+{
+	network_sortsupport_state *uss = ssup->ssup_extra;
+	inet	   *authoritative = DatumGetInetPP(original);
+	Datum		res;
+	Datum		ipaddr_int,
+				netmask_int,
+				subnet_int;
+	char		datum_size_left,
```

Comment:
Make these int -- there will be a promotion to int for shifts and stuff anyway. Don't be parsimonious with stack space if it's a single int v char or something.

Reply:
Changed to `uint8`.